jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java

```diff
@@ -788,7 +792,158 @@ protected void cleanupAfterStop() throws Exception
 
         // if we're not persisting the temp dir contents delete it
         if (tempDirectory != null && tempDirectory.exists() && !isTempDirectoryPersistent())
+        {
             IO.delete(tempDirectory);
+        }
+
+        //if it was jetty that created the tmp dir, it can be reset, otherwise we need to retain the name
+        if (_createdTempDirectoryName)
+        {
+            setTempDirectory(null);
+            _createdTempDirectoryName = false;
+        }
+    }
+
+    protected void makeTempDirectory()
+        throws Exception
+    {
+        File parent = getServer().getContext().getTempDirectory();
+        if (parent == null || !parent.exists() || !parent.canWrite() || !parent.isDirectory())
+            throw new IllegalStateException("Parent for temp dir not configured correctly: " + (parent == null ? "null" : "writeable=" + parent.canWrite()));
+
+        boolean persistent = isTempDirectoryPersistent() || "work".equals(parent.toPath().getFileName().toString());
+
+        //Create a name for the temp dir
+        String temp = getCanonicalNameForTmpDir();
+        File tmpDir;
+        if (persistent)
+        {
+            //if it is to be persisted, make sure it will be the same name
+            //by not using File.createTempFile, which appends random digits
+            tmpDir = new File(parent, temp);
+        }
+        else
+        {
+            // ensure dir will always be unique by having classlib generate random path name
+            tmpDir = Files.createTempDirectory(parent.toPath(), temp).toFile();
+            tmpDir.deleteOnExit();
+        }
+
+        if (LOG.isDebugEnabled())
+            LOG.debug("Set temp dir {}", tmpDir);
+        setTempDirectory(tmpDir);
+        setTempDirectoryPersistent(persistent);
+        _createdTempDirectoryName = true;
```

Comment:
if somebody subsequently calls setTempDirectory, then this needs to be set back to false.

Reply:
I reset it during `doStop()`, but you're saying you want every call to `setTempDirectory()` to unset it?

Reply:
Yes. Because if somebody calls set after make, then they are taking ownership. If the value they set is different, perhaps you should delete in set of created is true.

Reply:
done.